---
title: "GUIDEseq user's guide"
bibliography: references.bib
author:
- name: Lihua Julie Zhu
  affiliation: University of Massachusetts Chan Medical School
- name: Michael Lawrence
  affiliation: University of Massachusetts Chan Medical School
- name: Ankit Gupta
  affiliation: University of Massachusetts Chan Medical School
- name: Herve Pages
  affiliation: University of Massachusetts Chan Medical School
- name: Alper Kucukural
  affiliation: University of Massachusetts Chan Medical School
- name: Manuel Garber
  affiliation: University of Massachusetts Chan Medical School
- name: Scot Wolfe
  affiliation: University of Massachusetts Chan Medical School
output:
  BiocStyle::html_document:
    toc_float: true
  BiocStyle::pdf_document:
package: GUIDEseq
abstract: |
  Instructions on using GUIDEseq in analyzing GUIDEseq data.
vignette: |
  %\VignetteIndexEntry{Authoring R Markdown vignettes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r echo = FALSE}
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff=80),tidy=TRUE)

```

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

```{r kai_macro, echo=FALSE}
Rfunc <- function(funcname)
    sprintf('<span style="color:rgb(224,118,56);background-color:rgb(240,240,240)">%s</span>', funcname)
```

# Introduction
The most recently developed genome editing system, CRISPR-Cas9 has greater inherent flexibility than prior programmable nuclease platforms because sequence-specific recognition resides primarily within the associated sgRNA, which permits a simple alteration of its recognition sequence. The short Protospacer Adjacent Motif (PAM), which is recognized by Cas9, is the chief constraint on the target site design density. Because of its simplicity and efficacy, this technology is revolutionizing biological studies and holds tremendous promise for therapeutic applications [@RN647; @RN632].  

However, imperfect cleavage specificity of CRISPR/Cas9 nuclease within the genome is a cause for concern for its therapeutic application. S. pyogenes Cas9 (SpyCas9)-based nucleases can cleave an imperfect heteroduplex formed between the guide sequence and a DNA sequence containing a functional PAM where the number, position and type of base mismatches can impact the level of activity [@RN652; @RN638; @RN633]. This degree of promiscuity is problematic for therapeutic applications, since the generation of DNA breaks at unintended (off-target) sites has the potential to alter gene expression and function through direct mutagenesis or the generation of genomic rearrangements. Experimentally defining the number and activity of off-target sites for a given Cas9/sgRNA complex genome-wide is critical to assess and improve nuclease precision.

A new suite of genome-wide off-target detection methods have recently been described that can identify sites with low cleavage activity within a population of nuclease-treated cells. One of the most sensitive and straightforward methods to employ is GUIDE-seq [@RN653]. This method relies on erroneous NHEJ-mediated DNA repair to capture co-introduced blunt-ended double stranded oligonucleotides (dsODNs) at Cas9-induced breakpoints within the genome. The GUIDE-seq dsODNs display high insertion frequency (up to 50% of the measured indel rate [@RN653]) at Cas9-induced DSBs, thereby tagging these loci for selective amplification and subsequent deep sequencing. The method is quite sensitive as off-target sites with >0.1% indel frequency can be detected, and the frequency of dsODN insertion appears to be correlated with the frequency of Cas9-induced lesions at each site [@RN653]. This method has been used successfully to evaluate the precision of Cas9 and its variants (tru-sgRNAs [@RN653] or PAM variants [@RN646]). Given its favorable properties, GUIDE-seq could become a standard in the nuclease field for off-target analysis.

While the GUIDE-seq method is straightforward to employ, to date no bioinformatic tools have been released to the community to support the analysis of this data. We developed `r Biocpkg('GUIDEseq')` package [@RN656] to facilitate the analysis of GUIDE-seq dataset, including retaining one read per unique molecular identifier (UMI), filtering reads lacking integration oligo sequence (dsODNs), identifying peak locations (cleavage sites) and heights, merging cleavage sites from plus strand and those from minus strand, and performing target and off target search of the input gRNA. This analysis leverages our `r Biocpkg('ChIPPeakAnno')` package [@RN654] for merging cleavage sites from plus strand and minus strand, and `r Biocpkg('CRISPRseek')` package [@RN655] for defining the homology of any identified off-target site to the guide sequence and Cas9 PAM specificity. 

# Workflow of GUIDE-seq data analysis
Here is the workflow of GUIDE-seq data analysis with human sequence. First load `r Biocpkg('GUIDEseq')` and `r Biocannopkg('BSgenome.Hsapiens.UCSC.hg19')`.

To find BSgenome of other species, please refer to available.genomes in the
`r Biocpkg('BSgenome')` package. For example, 
`r Biocannopkg('BSgenome.Hsapiens.UCSC.hg19')` for hg19, 
`r Biocannopkg('BSgenome.Mmusculus.UCSC.mm10')` for mm10, 
`r Biocannopkg('BSgenome.Celegans.UCSC.ce6')` for ce6, 
`r Biocannopkg('BSgenome.Rnorvegicus.UCSC.rn5')` for rn5,
`r Biocannopkg('BSgenome.Drerio.UCSC.danRer7')` for Zv9, 
and `r Biocannopkg('BSgenome.Dmelanogaster.UCSC.dm3')` for dm3

Then specify the alignment file path as alignment.inputfile, and a umi file path as umi.inputfile containing unique molecularidentifier for each sequence.

```{r echo=TRUE,message=FALSE,warning=FALSE}
library(GUIDEseq)
umifile <- system.file("extdata", "UMI-HEK293_site4_chr13.txt", package = "GUIDEseq")
bedfile <- system.file("extdata","bowtie2.HEK293_site4_chr13.sort.bed", package = "GUIDEseq")
bamfile <- system.file("extdata","bowtie2.HEK293_site4_chr13.sort.bam", package = "GUIDEseq")

```

The alignment.inputfile is an alignment file in bed format containing CIGAR information. The alignment.inputfile contains chromosome, start, end, readID, mapping quality, strand and CIGAR information as a tab delimited file. Here is an example line.
```
chr13 27629253 27629403 HWI-M01326:156:1:113:4572:6938/1 44 + 150M
```
Scripts for bin reads, remove adaptor, mapping to genome are available at [GUIDE-seq](http://mccb.umassmed.edu/GUIDE-seq/). 

The umi.inputfile is a tab delimited file containing at least two columns, read IDs and corresponding unique molecular identifiers (UMI). Script for creating umi.inputfile is available as [_getUmi.pl_](http://mccb.umassmed.edu/GUIDE-seq/getUmi.pl). An example input file is at [testGetUmi](http://mccb.umassmed.edu/GUIDE-seq/testGetUmi/). Please make sure to use R1 reads as input to _getUmi.pl_.

## Step 1: Remove PCR bias and obtain unique cleavage events 
PCR amplification often leads to biased representation of the starting sequence population. To track the sequence tags present in the initial sequence library,unique molecular identifiers (UMI) are added to the 5 prime of each sequence in the staring library. The function `r Rfunc("getUniqueCleavageEvents")` uses the UMI sequence in the umi.inputfile (optionally contains umi plus the first few sequence from R1 reads) to obtain the starting sequence library. It also filters out reads that does not contain the integration oligo sequence, too short or not in the right paired configuration. 

For detailed parameter settings for function `r Rfunc("getUniqueCleavageEvents")`, please type `help(getUniqueCleavageEvents)` or `?getUniqueCleavageEvents`.

```{r echo=TRUE,message=FALSE,warning=FALSE}
uniqueCleavageEvents <- getUniqueCleavageEvents(bamfile, umifile, n.cores.max = 1)
uniqueCleavageEvents$cleavage.gr

```

## Step 2: Summarize cleavage events
Calling the function `r Rfunc("getPeaks")` with the results from `r Rfunc("getUniqueCleavageEvents")` outputs summarized cleavage events for each moving window with at least `min.reads` of cleavage events.

By default, `window.size` is set to 20, `step` is set to 20, and `min.reads` is set to 2. For detailed parameter settings using function `r Rfunc("getPeaks")`, please type `help(getPeaks)`.

```{r echo=TRUE,message=FALSE,warning=FALSE}
peaks <- getPeaks(uniqueCleavageEvents$cleavage.gr, min.reads = 80)
peaks.gr <- peaks$peaks
peaks.gr

```

## Step 3: Merge peaks from plus and minus strand
Calling the function `r Rfunc("mergePlusMinusPeaks")` with the output from `r Rfunc("getPeaks")` to merge peaks from plus strand and minus strand with specific orientation and within certain distance apart.

By default, `plus.strand.start.gt.minus.strand.end` is set to TRUE and `distance`.threshold is set to 40, i.e., twice of the `window.size`. For detailed parameter settings using function `r Rfunc("mergePlusMinusPeaks")`, please type `help(mergePlusMinusPeaks)`.

```{r echo=TRUE,message=FALSE,warning=FALSE}
mergedPeaks <- mergePlusMinusPeaks(peaks.gr = peaks.gr, output.bedfile = "mergedPeaks.bed")
mergedPeaks$mergedPeaks.gr

head(mergedPeaks$mergedPeaks.bed)

```

## Step 4: Off target analysis of extended regions around the identified cleavage sites
Calling the function `r Rfunc("offTargetAnalysisOfPeakRegions")` with input gRNA, peaks and genome of interest, to annotate identified cleavage sites with sequence homolgy to input gRNA. For detailed parameter settings using function `r Rfunc("offTargetAnalysisOfPeakRegions")`. For more information, please type `help(offTargetAnalysisOfPeakRegions)`.

```{r echo=TRUE,message=FALSE,warning=FALSE,results="hide"}
library(BSgenome.Hsapiens.UCSC.hg19)
peaks <- "mergedPeaks.bed" # from step 3
gRNAs <- system.file("extdata","gRNA.fa", package = "GUIDEseq")
outputDir <- getwd()

offTargets <- offTargetAnalysisOfPeakRegions(
						gRNA = gRNAs,
						peaks = peaks,
						format=c("fasta", "bed"),
						peaks.withHeader = FALSE,
						BSgenomeName = Hsapiens,
						outputDir = outputDir,
						orderOfftargetsBy = "predicted_cleavage_score",
						allowed.mismatch.PAM = 2,
						overwrite = TRUE,
						upstream = 50, 
						downstream = 50, 
						PAM.size = 3, 
						gRNA.size = 20,
						PAM = "NGG",
						PAM.pattern = "(NAG|NGG|NGA)$", 
						max.mismatch = 8)
```
```{r}
head(offTargets)

```

Another example:

```{r echo=TRUE,message=FALSE,warning=FALSE,results="hide"}
peaks <- system.file("extdata", "T2plus100OffTargets.bed", package = "CRISPRseek")
gRNAs <- system.file("extdata", "T2.fa", package = "CRISPRseek")
offTargets <- offTargetAnalysisOfPeakRegions(
						gRNA = gRNAs, 
						peaks = peaks,
						format=c("fasta", "bed"),
						peaks.withHeader = TRUE, 
						BSgenomeName = Hsapiens,
						outputDir = outputDir,
						orderOfftargetsBy = "predicted_cleavage_score",
						allowed.mismatch.PAM = 2, 
						overwrite = TRUE,
						upstream = 50, 
						downstream = 50, 
						PAM.size = 3, 
						gRNA.size = 20,
						PAM = "NGG",
						PAM.pattern = "(NAG|NGG|NGA)$", 
						max.mismatch = 8)

```
```{r}
head(offTargets)

```
Note that in the above example, `peaks.withHeader` is set to TRUE because `T2plus100OffTargets.bed` contains a header line. In contrast, the `mergedPeaks.bed` file created with step3 doesn't contain a header line and `peaks.withHeader` was set to FALSE accordingly.

## Run all steps in one workflow function
The function `r Rfunc("GUIDEseqAnalysis")` is a wrapper function that uses the UMI sequence or plus the first few bases of each sequence from R1 reads to estimate the starting sequence library, piles up reads with a user defined window and step size, identify the cleavage sites, merge cleavage sites from plus strand and minus strand, followed by off target analysis of extended regions around the identified cleavage sites. For detailed parameter settings using function `r Rfunc("GUIDEseqAnalysis")`, please type `help(GUIDEseqAnalysis)`.

```{r echo=TRUE,message=FALSE,warning=FALSE,results=FALSE}
gRNA.file <- system.file("extdata", "gRNA.fa", package = "GUIDEseq")
system.time(guideSeqRes <- GUIDEseqAnalysis(
													alignment.inputfile = bamfile, 
													umi.inputfile = umifile, 
													gRNA.file = gRNA.file, 
													orderOfftargetsBy = "peak_score",
													descending = TRUE, 
													n.cores.max = 1,
													BSgenomeName = Hsapiens, 
													min.reads = 1))

```
```{r}
names(guideSeqRes)

```

# References
<div id="refs"></div>

# Session info
```{r sessionInfo, echo=TRUE}
sessionInfo()

```